Enum conversion not working: IllegalArgumentException: argument type mismatch for step argument

## Summary

Regression between 3.0.2 and 2.4.0 breaks converting enum parameters of steps


## Expected Behavior

After reverting to 2.4.0:

```
> Task :cucumber
Feature: Calculations

  Scenario: Divide              # src/test/resources/calculation.feature:2
Doing Division
    When Division is calculated # CalculationStepdefs.splitInfoPfShiftIsSelected(CalculationType)

1 Scenarios (1 passed)
1 Steps (1 passed)
```

## Current Behavior

~~~
> Task :cucumber FAILED
Feature: Calculations

  Scenario: Divide              # src/test/resources/calculation.feature:2
    When Division is calculated # CalculationStepdefs.splitInfoPfShiftIsSelected(CalculationType)
      cucumber.runtime.CucumberException: Failed to invoke com.example.CalculationStepdefs.splitInfoPfShiftIsSelected(CalculationType) in file:/home/piotr-kubowicz/temp/bugs/cucumber3/build/classes/java/test/, caused by java.lang.IllegalArgumentException: argument type mismatch                                                                                                                            
        at cucumber.runtime.Utils$1.call(Utils.java:29)                                                                               
        at cucumber.runtime.Timeout.timeout(Timeout.java:16)                                                                          
        at cucumber.runtime.Utils.invoke(Utils.java:20)                                                                               
        at cucumber.runtime.java.JavaStepDefinition.execute(JavaStepDefinition.java:48)                                               
        at cucumber.runtime.PickleStepDefinitionMatch.runStep(PickleStepDefinitionMatch.java:50)                                      
        at cucumber.runner.TestStep.executeStep(TestStep.java:55)                                                                     
        at cucumber.runner.TestStep.run(TestStep.java:42)                                                                             
        at cucumber.runner.PickleStepTestStep.run(PickleStepTestStep.java:53)                                                         
        at cucumber.runner.TestCase.run(TestCase.java:47)                                                                             
        at cucumber.runner.Runner.runPickle(Runner.java:44)                                                                           
        at cucumber.runtime.Runtime.runFeature(Runtime.java:120)                                                                      
        at cucumber.runtime.Runtime.run(Runtime.java:106)                                                                             
        at cucumber.api.cli.Main.run(Main.java:35)                                                                                    
        at cucumber.api.cli.Main.main(Main.java:18)                                                                                   
Caused by: java.lang.IllegalArgumentException: argument type mismatch                                                                 
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)                                                                
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)                                              
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                                      
        at java.lang.reflect.Method.invoke(Method.java:498)                                                                           
        at cucumber.runtime.Utils$1.call(Utils.java:26)                                                                               
        ... 13 more                                                                                                                   
~~~

## Possible Solution

<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->

## Steps to Reproduce (for bugs)

src/test/resources/calculation.feature
```
Feature: Calculations
  Scenario: Divide
    When Division is calculated
```

```
package com.example;

import cucumber.api.java.en.When;

public class CalculationStepdefs {

    @When("([\\S]+) is calculated")
    public void splitInfoPfShiftIsSelected(CalculationType calculationType) {
        System.out.println("Doing " + calculationType);
    }
}
```

```
package com.example;

public enum CalculationType {
    Multiplication,
    Division
}
```

build.gradle
```
plugins { id 'java-library' }

def cucumberVersion = '3.0.2'
dependencies {
    testImplementation "io.cucumber:cucumber-java8:$cucumberVersion"
}

repositories {
    jcenter()
}

task cucumber() {
    dependsOn testClasses
    doLast {
        javaexec {
            main = "cucumber.api.cli.Main"
            classpath = configurations.testRuntimeClasspath + sourceSets.main.output + sourceSets.test.output
            args = ['--plugin', 'pretty', '--glue', 'com.example', 'src/test/resources']
        }
    }
}
build.dependsOn cucumber
```

## Context & Motivation

This blocks me from upgrading to 3.0.2.

## Your Environment

Java(TM) SE Runtime Environment (build 1.8.0_172-b11)
Ubuntu 17.10

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,8 @@ Please see [CONTRIBUTING.md](https://github.com/cucumber/cucumber/blob/master/CO
  * [Core] Add --wip option ([#1381](https://github.com/cucumber/cucumber-jvm/pull/1381) Heziode)
  * [Core] Upgrade gherkin to 5.1.0  ([#1377](https://github.com/cucumber/cucumber-jvm/pull/1377) Aslak Hellesøy)
  * [Weld] Document the need for a beans.xml per source root ([#923](https://github.com/cucumber/cucumber-jvm/pull/923) Harald Albers)
+ * [Core] Upgrade cucumber-expressions to [6.1.0](https://github.com/cucumber/cucumber/blob/master/cucumber-expressions/CHANGELOG.md#610---2018-09-23) ([#1464](https://github.com/cucumber/cucumber-jvm/pull/1464) M.P. Korstanje)
+    * ParameterType.fromEnum(MyEnumClass.class) to make it easier to register enums.
  
 ### Changed
   * [Core] Upgrade datatable to 1.1.3 ([#1414](https://github.com/cucumber/cucumber-jvm/pull/1414) Łukasz Suski)

--- a/core/src/test/java/cucumber/runtime/formatter/PrettyFormatterTest.java
+++ b/core/src/test/java/cucumber/runtime/formatter/PrettyFormatterTest.java
@@ -409,7 +409,7 @@ public class PrettyFormatterTest {
 
         TypeRegistry registry = new TypeRegistry(Locale.ENGLISH);
         StepExpressionFactory stepExpressionFactory = new StepExpressionFactory(registry);
-        StepExpression expression = stepExpressionFactory.createExpression("the order is placed( and (not yet )?confirmed)?");
+        StepExpression expression = stepExpressionFactory.createExpression("^the order is placed( and (not yet )?confirmed)?$");
 
         PrettyFormatter prettyFormatter = new PrettyFormatter(null);
         String stepText = "the order is placed and not yet confirmed";
@@ -427,7 +427,7 @@ public class PrettyFormatterTest {
         PrettyFormatter prettyFormatter = new PrettyFormatter(null);
         TypeRegistry registry = new TypeRegistry(Locale.ENGLISH);
         StepExpressionFactory stepExpressionFactory = new StepExpressionFactory(registry);
-        StepExpression expression = stepExpressionFactory.createExpression("the order is placed( and (not( yet)? )?confirmed)?");
+        StepExpression expression = stepExpressionFactory.createExpression("^the order is placed( and (not( yet)? )?confirmed)?$");
         String stepText = "the order is placed and not yet confirmed";
         String formattedText = prettyFormatter.formatStepText("Given ", stepText, formats.get("passed"), formats.get("passed_arg"), DefinitionArgument.createArguments(expression.match(stepText)));
 

--- a/pom.xml
+++ b/pom.xml
@@ -82,7 +82,7 @@
         <xmlunit.version>1.6</xmlunit.version>
         <tag-expressions.version>1.1.1</tag-expressions.version>
         <typetools.version>0.5.0</typetools.version>
-        <cucumber-expressions.version>6.0.1</cucumber-expressions.version>
+        <cucumber-expressions.version>6.1.0</cucumber-expressions.version>
         <datatable.version>1.1.3</datatable.version>
     </properties>
     <licenses>

